pkg/safeio/safeio.go

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+// Copyright 2009 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package safeio
+
+import (
+	"fmt"
+	"io"
+)
+
+// ErrLimitReached indicates that ReadAllLimit has
+// reached its limit before completing a full read
+// of the io.Reader.
+var ErrLimitReached = fmt.Errorf("read limit reached")
+
+// ByteSize expresses the size of bytes
+type ByteSize float64
+
+const (
+	_ = iota // ignore first value by assigning to blank identifier
+	// KB is a Kilobyte
+	KB ByteSize = 1 << (10 * iota)
+	// MB is a Megabyte
+	MB
+	// GB is a Gigabyte
+	GB
+	// TB is a Terabyte
+	TB
+	// PB is a Petabyte
+	PB
+	// EB is an Exabyte
+	EB
+	// ZB is a Zettabyte
+	ZB
+	// YB is a Yottabyte
+	YB
+)
+
+// String converts a ByteSize to a string
+func (b ByteSize) String() string {
+	switch {
+	case b >= YB:
+		return fmt.Sprintf("%.1fYB", b/YB)
+	case b >= ZB:
+		return fmt.Sprintf("%.1fZB", b/ZB)
+	case b >= EB:
+		return fmt.Sprintf("%.1fEB", b/EB)
+	case b >= PB:
+		return fmt.Sprintf("%.1fPB", b/PB)
+	case b >= TB:
+		return fmt.Sprintf("%.1fTB", b/TB)
+	case b >= GB:
+		return fmt.Sprintf("%.1fGB", b/GB)
+	case b >= MB:
+		return fmt.Sprintf("%.1fMB", b/MB)
+	case b >= KB:
+		return fmt.Sprintf("%.1fKB", b/KB)
+	}
+	return fmt.Sprintf("%.1fB", b)
+}
+
+// ReadAllLimit reads from r until an error, EOF, or after n bytes and returns
+// the data it read. A successful call returns err == nil, not err == EOF.
+// Because ReadAllLimit is defined to read from src until EOF it does not
+// treat an EOF from Read as an error to be reported. If the limit is reached
+// ReadAllLimit will return ErrLimitReached as an error.
+func ReadAllLimit(r io.Reader, n ByteSize) ([]byte, error) {
```

Comment:
This seems like a duplication of a [LimitReader](https://pkg.go.dev/io#LimitReader). Could we use that instead?

Reply:
Are you saying that we could do something like this?

```golang
io.ReadAll(io.LimitReader(r, size))
```

If so, I did investigate that as a solution and it has a problem. `io.LimitedReader` returns `io.EOF` when it hits the specified limit. Without counting bytes, it is actually impossible to know when `io.ReadAll`  (or `io.Read`) is invoked whether or not a returned `io.EOF` means that limit specified was reached or if the reader happens to be the same length as the limit. 

In order to mimic the same return profile of `io.ReadAll` (wherein `io.EOF` is not returned, as it is expected) we need to count bytes. Hitting the limit needs to be reported as an error.

Reply:
I was thinking along the lines of the following: (untested)

```go
buf, err := io.ReadAll(io.LimitReader(r, size))
if err != nil {
  return buf, err
}
if n, err := r.Read(make([]byte, 1)); err == io.EOF && n == 0 {
  return buf, nil
}
return buf, ErrLimitReached
```

That is, trying to avoid copying ReadAll by issuing an extra Read for a single byte. In terms of the extra `Read` that we issue, I don't think we need to be optimizing for the case of exceeding the limit, as that is not the expected case, probably?

FWIW, I've now found https://github.com/golang/go/issues/51115, where a very similar discussion took place.

Reply:
> I was thinking along the lines of the following: (untested)

If you can prove to me that every implementation of `io.Reader` returns `io.EOF` on multiple reads past the "real" end of whatever underlies `io.Reader`, then I would feel comfortable with this approach. I agree in theory that this is better, but I dislike the idea of trusting that everyone's implementation of `io.Reader` will do the "right thing".

Reply:
> returns io.EOF on multiple reads past the "real" end of whatever underlies io.Reader

Sure, fair point, in the case where the underlying data is exactly size, we rely on `Read` to return EOF repeatedly. I obviously cannot prove that.

We could just do this instead, anyway:

```go
buf, err := io.ReadAll(io.LimitReader(r, size+1))
if err != nil {
  return buf, err
}
if len(buf) > size {
  return buf[:size], ErrLimitReached
} else {
  return buf, nil
}
```

Reply:
Oh, very clever. I wish I had thought of that.